2023/Q1/artifacts/PoS-pseudocode/PoS-model-redelegation.md

```diff
@@ -222,19 +275,63 @@ tx_withdraw_unbonds_delegator(delegator_address)
 }
 ```
 
-## PoS functions
+The transaction `tx_redelegate` creates a redelegation from source validator `src_validator_address` to `dest_validator_address` for any tokens that a delegator `delegator_address` has currently delegated to the source validator. The transactions takes the following steps:
 
+1. The function first checks that the source validator is not frozen. This is to prevent unbonding at an epoch `e` from a validator that it is known to have behaved, but the slash has not been processed yet.
+2. Then it checks that the redelegation is not a chain redelegation, in which case it prevents it by returning.
+   - The design does not track chain redelegation precisely.
+   - Instead approximates it by preventing a validator to redelegate tokens belonging to a delegator if it has recently received a redelegation of tokens belonging to the same delegator.
+   - The `inconming_redelegations` variable is used to implement this mechanism.
+3. The function computes the total amount of tokens that the delegator has bonded at the source validator (`bonded_tokens`).
```

Comment:
It should be possible to only redelegate a portion of the tokens that the delegator originally had delegated to the source validator, as well, right?

Reply:
Yes, this would be nice (at some point, it's not urgent)